python-package/lightgbm/callback.py

```diff
@@ -1,10 +1,15 @@
 # coding: utf-8
 """Callbacks library."""
 import collections
+from dataclasses import dataclass
 from functools import partial
-from typing import Any, Callable, Dict, List, Tuple, Union
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Tuple, Union
+
+from .basic import Booster, _ConfigAliases, _LGBM_BoosterEvalMethodResultType, _log_info, _log_warning
+
+if TYPE_CHECKING:
+    from .engine import CVBooster
```

Comment:
See https://mypy.readthedocs.io/en/stable/runtime_troubles.html#import-cycles for an explanation of this.

Adding this pattern to avoid cyclical imports... `engine.CVBooster` is needed for the type annotation on `CallbackEnv.model`, but `engine.py` also tries to import all of `callback`'s contents.

https://github.com/microsoft/LightGBM/blob/709ea4cad3ff4c8834bdf389eef8016345d9a1ac/python-package/lightgbm/engine.py#L12